src/ucp/rndv/rndv.c

```diff
@@ -725,47 +816,45 @@ static ucs_status_t ucp_rndv_req_send_rma_get(ucp_request_t *rndv_req,
 UCS_PROFILE_FUNC_VOID(ucp_rndv_recv_frag_put_completion, (self),
                       uct_completion_t *self)
 {
-    ucp_request_t *freq              = ucs_container_of(self, ucp_request_t,
-                                                        send.state.uct_comp);
-    ucp_worker_h worker              = freq->send.ep->worker;
-    ucs_ptr_map_key_t rreq_remote_id = freq->send.rndv_put.rreq_remote_id;
-    int is_put_proto                 = (rreq_remote_id == UCP_REQUEST_ID_INVALID);
-    ucp_request_t *req               = freq->super_req;
-    ucp_request_t *rndv_req          = NULL;
-
-    ucs_trace_req("freq:%p: recv_frag_put done. rreq:%p ", freq, req);
+    ucp_request_t *freq     = ucs_container_of(self, ucp_request_t,
+                                               send.state.uct_comp);
+    int is_put_proto        = (freq->super_req->flags &
+                               (UCP_REQUEST_FLAG_RECV_TAG |
+                                UCP_REQUEST_FLAG_RECV_AM));
+    ucp_request_t *rreq     = NULL;
+    ucp_request_t *rndv_req = NULL;
 
     /* release memory descriptor */
-    ucs_mpool_put_inline((void *)freq->send.mdesc);
+    ucs_mpool_put_inline((void*)freq->send.mdesc);
 
     /* rndv_req is NULL in case of put protocol */
     if (!is_put_proto) {
-        /* it is local operation, expected that a request will be always valid */
-        rndv_req = ucp_worker_get_request_by_id(worker, rreq_remote_id);
-        ucs_assert(rndv_req != NULL);
+        ucs_trace_req("freq:%p: recv_frag_put done, nrdv_req:%p rreq:%p ", freq,
+                      rndv_req, rreq);
+
+        rndv_req = freq->super_req;
+        rreq     = rndv_req->super_req;
 
         /* pipeline recv get protocol */
         rndv_req->send.state.dt.offset += freq->send.length;
 
         /* send ATS for fragment get rndv completion */
         if (rndv_req->send.length == rndv_req->send.state.dt.offset) {
-            ucp_rkey_destroy(rndv_req->send.rndv_get.rkey);
-            ucp_rndv_req_send_ats(rndv_req, req,
-                                  rndv_req->send.rndv_get.remote_req_id,
-                                  UCS_OK);
+            ucp_rkey_destroy(rndv_req->send.rndv.rkey);
+            ucp_rndv_req_send_ats(rndv_req, rreq,
+                                  rndv_req->send.rndv.remote_req_id, UCS_OK);
         }
+    } else {
+        ucs_trace_req("freq:%p: recv_frag_put done, rreq:%p ", freq, rreq);
+        rreq = freq->super_req;
     }
 
-    ucs_assertv(req->recv.remaining >= freq->send.length,
-                "req->recv.remaining %zu, freq->send.length %zu",
-                req->recv.remaining, freq->send.length);
-    req->recv.remaining -= freq->send.length;
-    if (req->recv.remaining == 0) {
-        ucp_rndv_recv_req_complete(req, UCS_OK);
-        if (!is_put_proto) {
-            ucs_assert(rndv_req != NULL);
-            ucp_worker_del_request_id(worker, rndv_req, rreq_remote_id);
```

Comment:
seems this line is removed. is it not needed?

Reply:
yes, it is not needed anymore
I've handled it a different way - in `ucp_rndv_recv_frag_put_completion()` we're relying on flags set in `super_req`:
- they are not set to RECV_AM/RECV_TAG for `rndv_req`
- they are set to RECV_AM/RECV_TAG for `rreq`
i.e. now we don't overwrite `freq->super_req` by `rreq`. So, in case of GET protocol, it is `freq` -> `rndv_req` -> `rreq`. For PUT protocol - `freq` -> `rreq`